File: Python/Combination Sum II.py
class Solution:
    # @param candidates, a list of integers
    # @param target, integer
    # @return a list of lists of integers
    def __init__(self):
        ret = []
        
    def combinationSum2(self, candidates, target):
        ret = []
        candidates.sort() 
        for i,item in enumerate(candidates):
            if target>item:
                subSet=self.combinationSum2(candidates[i+1:],target-item) 
                if subSet:
                    for lis in subSet:
                        if not [item] + lis in ret:
                            ret += [[item] + lis]

            elif target==item:
                newSet = set()
                if not [item] in ret: 
                    ret+=[[item]]

            else:
                break
        return ret

File: Java/src/Combination Sum II.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Solution {
    public List<List<Integer>> combinationSum2(int[] num, int target) {
      List<List<Integer>> ret = new ArrayList<>();
        Arrays.sort(num);
        for(int i = 0; i<num.length; i++){
            if (num[i] < target){
                 List<List<Integer>> subSet = combinationSum2(Arrays.copyOfRange(num,i+1,num.length), target- num[i]);
                 
                 for(List<Integer> list : subSet){
                    list.add(0,num[i]);
                    if (!ret.contains(list)){
                         ret.add(list);
                    }
                   
                 }
            }
            else if (num[i] == target){
                List<Integer> newList = new ArrayList<Integer>();
                newList.add(num[i]);
                if (!ret.contains(newList)){
                        ret.add(newList);
                    }
                
            }
            else{
                break;
            }
        }
        return ret;
    }
}